### DISCUSSION TITLE
How to assign static port to TCp server clients

### DISCUSSION THREAD

**Discussion by nirajgenius** [jchristn/WatsonTcp#141] [Q&A]
Hello Joel,

How can we make sure when the client gets connected then it uses the static port.
When I say static means once the client first time joined with the server then always assigned that same port one future connections.
Please see my screenshot below. I have the same client connected two times and each time It's assigned a different port.
![2](https://user-images.githubusercontent.com/11919877/120270329-1023e380-c2c7-11eb-8857-9bacd1f7b6b3.jpg)
![1](https://user-images.githubusercontent.com/11919877/120270332-11551080-c2c7-11eb-9415-9c346c00a378.jpg)

**Comment by jchristn** [comment]
Hi @nirajgenius thanks for your note.  

The library doesn't support that today, but the underlying ```TcpClient``` does: https://stackoverflow.com/questions/2869840/is-there-a-way-to-specify-the-local-port-to-used-in-tcpclient

Note that I don't know if there are multiplatform implications, i.e. it works on Windows but not on Mac.  I haven't tested it.

Out of curiosity, why would you want to make the client port static?  Generally by definition a client port is random, above 1024, and not something you should want to persist.

Please let me know more about the use case if you have some time; there might be an easy way to incorporate this into WatsonTcp without disturbing the codebase too much, but I want to make sure it's inline with the intent of the library.

Cheers, Joel

**Reply by jchristn** [reply]
Actually, give this a try and let me know if it works for you.

NuGet v4.8.6: https://www.nuget.org/packages/WatsonTcp/4.8.6
Commit: https://github.com/jchristn/WatsonTcp/commit/23fd24293ef94a8b6989f7a91bb4ac9fc4cae543

Usage:
```
client.Settings.LocalPort = 5000;
```

Note: a ```SocketException``` will be thrown when you call ```Connect()``` if the port number is already in use.

**Comment by nirajgenius** [comment]
Hello Joel,

Thank you for your reply.

We have Serial to WiFi ethernet Wireless devices that will be considered as clients and by using the Watson  TCP library I build a server as you saw in the above screenshot. So far I am able to connect multiple clients but the problem is how to identify who is which client?

how to make sure if I want to communicate to a specific client?
I thought if I assign a specific dedicate port to each client then I can identify which client to send message Or Command.

I was thinking IP address but in my case, the client can be multiple in the same network so they will have the same IP so IP: PORT is what I think we can identify the unique client.

can you suggest how to identify clients?

**Reply by jchristn** [reply]
Got it.

You can either use the version I specified above, and maintain a port<>identifier mapping on your side, or (probably much better and supporting of a dynamic environment) would be to package metadata with your messages.

```
Dictionary<string, string> md = new Dictionary<string, string>();
md.Add("Identifier", "MyNodeName");
client.Send("hello world", md);
```

On the server side, you will also have this object available to you in ```MessageReceivedEventArgs.Metadata```.

**Reply by nirajgenius** [reply]
Thank you Joel for your help.
Let me implement the way you specified above using "Identifier" metadata.

**Reply by nirajgenius** [reply]
[Logs.txt](https://github.com/jchristn/WatsonTcp/files/6591860/Logs.txt)
Hi Joel,

I have two different problems that I am facing with Super Simple TCP and Watson TCP.let me explain below

When I use Super simple TCP nugget package then I can get a response from the client when the client and servers are in the same intranet but when I deploy the console app to our public server and client and server connect with public IP then the client gets disconnected and connected many times and cont. that is happening so not sure what settings I need to do with the server.

as A option I am trying to use Watson TCP then It allows me cont. connection with client and server that resolve above problem but I am not able to get a response from clients when sending the same command from the server.

Could you please help me in this?

I have attached log file here with this message and also I have replaced clients Public IP with X.X.X.X

**Reply by nirajgenius** [reply]
A have attached log file above

**Reply by nirajgenius** [reply]
Hi Joel..

More on this..
I have set 
_Server.Keepalive.TcpKeepAliveInterval = 3600;
so super simple TCP is not disconnecting and reconnecting client continuously so that resolve my issue with Super simple TCP now I can keep both clients live for log time.

What is the default value of this property?
is it safe to set some high value like I set 3600?

I have downloaded your example for super simple TCP server project but that property is not set.
I have seen watson tcp and it was set so i used 10 as value and then used 3600.

**Reply by jchristn** [reply]
I would disable TCP keepalives on both the client and the server.  They're outside of my control (handled in the kernel) and have proven to be unreliable.

**Reply by nirajgenius** [reply]
I am just curious to know why my client is connected and disconnecting when it connects with Live Ip?

the log thats says like this : null header data, peer disconnect detected

Something related to header?

**Reply by jchristn** [reply]
There could be a million reasons, honestly.  It could be an intermittent disconnection on some intermediary router, too many dropped packets, packets routed across different networks and delivered out of an order and some upstream device thinks it's failed.  The message header doesn't have anything to do with it though, it's just payload data to TCP.  The ```null header data``` is because the call to read from the socket failed, thus it was unable to build the next message.  Cheers, Joel

**Reply by nirajgenius** [reply]
Thank you Joel,
Some more information to get it more clear.

So, Joel,I am using SimpleTcpServer class Or WatsonTcpServer Class to build server. using _Server.Start(); method and seting options.

but my clients are not using "WatsonTcpClient" class Or "SimpleTcpClient" class to send Or receive data.
in my case clients are hardware devices and they are USR ethernet to serial wifi module

in USR devices, we are just allowed to enter the TCP server IP..NO code can be done there. as soon as we enter the TCP server address they start finding the listening server and automatically connected.

Watson server when started then automatically connect it but it keeps connecting and disconnecting client when the connection made over internet public IP.it works fine when the client and server made the connection in the same intranet.

**Reply by jchristn** [reply]
Hi @nirajgenius so these devices in the field where you can't add code, I doubt they're going to be compatible with WatsonTcp, since WatsonTcp has a framing layer in the TCP payload (in between the TCP headers and your data).  

That 100% explains this in your logs:
```
[Debug    ] [WatsonTcpServer] accepted connection from X.X.X.X:1030
[Debug    ] [WatsonTcpServer] starting data receiver for X.X.X.X:1030
Client connected: X.X.X.X:1030
[Debug    ] [WatsonMessage] null header data, peer disconnect detected
[Debug    ] [WatsonTcpServer] disconnect detected for client X.X.X.X:1030
```
SimpleTcp however should work.  It does not have a framing layer.

I really couldn't tell you why they keep disconnecting.  The first step would be to get a network trace to see if there is a RST or FIN being sent in either direction.

**Reply by nirajgenius** [reply]
Yes, Joel, I can not do code on that client's wifi to serial modules...
Simple TCP works just fine when client and server are located in the intranet like in my office as an example.
same code when the server is built on my AWS production server then it keeps getting connected and disconnected.

So I tried different options and I get success to keep persistent connection by setting 
_Server.Keepalive.TcpKeepAliveInterval = 3600;
uncommented from your example code...

So far going well with Simple TCP...I am just afraid as you said in a previous comment before that the setting I mentioned above purely depends on kernel and proven to be unreliable.

Separately..in YAT(Yet Another Terminal)...I can build a server on the same production box and also multiple clients are connected to that server and they remain persistent that's why I am hopeful with the Simple TCP plugin and considering no network-related issues otherwise YAT should have faced it...

**Reply by jchristn** [reply]
Hi @nirajgenius yes I'd definitely **disable** keepalives.  I added them as part of a feature request (to support network cable removal, etc) and they tend to create problems with random disconnections.  Cheers, Joel
